tensorflow/core/kernels/dml_matmul_op.cc

```diff
@@ -331,17 +353,46 @@ class DmlBatchMatMulKernel : public DmlKernel {
         in1_physical_shape.dim_size(in1_physical_shape.dims() - 1));
 
     DmlKernelParams params;
-    params.kernel_input_indices = {
-        0, 1,
-        absl::nullopt  // We don't use GEMM's 'C' tensor
-    };
+    params.kernel_input_indices = {0, 1};
 
     DmlKernelTensors tensors = GetTensorInfos(ctx, params);
     tensors.inputs[0]->desc = CreateTensorDescFromInput(ctx, 0, in0_shape);
     tensors.inputs[1]->desc = CreateTensorDescFromInput(ctx, 1, in1_shape);
 
     auto input_descs = GetDmlTensorDescs(tensors.inputs);
-    auto output_descs = GetDmlTensorDescs(tensors.outputs);
+    auto scope = dml::Graph(ctx->GetDmlDevice());
+    auto a_tensor = dml::InputTensor(scope, 0, input_descs[0]);
+    auto b_tensor = dml::InputTensor(scope, 1, input_descs[1]);
+
+    auto a_tensor_sizes = a_tensor.GetOutputDesc().sizes;
+    auto b_tensor_sizes = b_tensor.GetOutputDesc().sizes;
+
+    // DML doesn't support more than 4 dimensions for GEMM, so do the
+    // broadcasting manually by using identity beforehand
+    if (a_tensor_sizes.size() > kNchwDimensionCount) {
+      a_tensor = dml::Identity(a_tensor);
```

Comment:
Can the identity copy be elided if the leading batch dimensions are all the same between in0 and in1? (like how we internally flatten leading dimensions in DML's `Optimize()` call for many operators)

Reply:
If all leading batch dimensions are the same, then `a_tensor_sizes.size()` will always be `3` because of the broadcasting done in the initialization helper.